src/components/tree/phyloTree/types.ts

```diff
@@ -0,0 +1,260 @@
+  }
+  children?: ReduxNode[]
+  currentGt?: number
+  
+  /** see the number of subtending tips (alive or dead) */
+  fullTipCount?: number
+
+  hasChildren?: boolean
+  inView?: boolean
+  name: string
+  node_attrs?: {
+    div?: number
+    hidden?: "always" | "timetree" | "divtree"
+    num_date?: {
+      value: number
+    }
+  }
+  parent?: ReduxNode
+  parentInfo?: {
+    original?: ReduxNode
+  }
+  shell?: PhyloNode
+
+  /** the number of visible tips */
+  tipCount?: number
+
+  unexplodedChildren?: ReduxNode[]
+}
+
+/** [root idx tree1, root idx tree2] */
+export type Root = [number | undefined, number | undefined]
+
+/** Properties can be any property on PhyloNode but as an array for multiple nodes */
+export interface PropsForPhyloNodes {
+  branchStroke?: string[]
+  fill?: string[]
+  r?: number[]
+  tipStroke?: (string | undefined)[]
+  visibility?: Visibility[]
+}
+
+// Workaround for CSSStyleDeclaration having strokeWidth instead of stroke-width.
+// https://github.com/microsoft/TypeScript/issues/17827#issuecomment-861847433
+interface CSSStyleDeclarationWithHyphens extends CSSStyleDeclaration {
+  "stroke-width": string
+}
+
+// Workaround for CSSStyleDeclaration having generic string types.
+// Potentially related issue: https://github.com/Microsoft/TypeScript/issues/17827
+// TODO: consider extending existing interfaces such as SVGCircleElement for cx/cy/r
+interface CustomCSSStyleDeclaration {
+  cx?: number
+  cy?: number
+  d?: any
+  r?: number
+  "stroke-width"?: number
+  visibility?: Visibility
+
+  // TODO: This should be `string | number`, conditional on layout
+  x: any
+
+  // TODO: This should be `string | number`, conditional on layout
+  y: any
+}
+
+// Apply fixes from above
+export type SVGProperty = Extract<keyof CSSStyleDeclarationWithHyphens | keyof CustomCSSStyleDeclaration, string>
+
+export interface PhyloNode extends Partial<Omit<CSSStyleDeclarationWithHyphens, keyof CustomCSSStyleDeclaration>>, CustomCSSStyleDeclaration {
+  angle?: number
+  branch?: [string, string]
+  branchStroke?: string
+  conf?: [number, number]
+
+  /** SVG path */
+  confLine?: string
+
+  crossDepth?: number
+  depth?: number
+  displayOrder?: number
+  displayOrderRange?: number[]
+  fill?: string
+  inView: boolean
+  n: ReduxNode
+  pDepth?: number
+  px?: number
+  py?: number
+  rot?: number
+  smallBigArc?: boolean
+  tau?: number
+  that: PhyloTree
+  tipStroke?: string
+  update?: boolean
+
+  /** SVG path */
+  vaccineCross?: string
+
+  w?: number
+  xBase?: number
+  xCBarEnd?: number
+  xCBarStart?: number
+  xCross?: number
+  xTip?: number
+  yBase?: number
+  yCBarEnd?: number
+  yCBarStart?: number
+  yCross?: number
+  yTip?: number
+}
+
+export interface ScatterVariables {
+  showBranches?: boolean
+  showRegression?: boolean
+  x?: string
+  xContinuous?: boolean
+  xDomain?: number[]
+  xTemporal?: boolean
+  y?: string
+  yContinuous?: boolean
+  yDomain?: number[]
+  yTemporal?: boolean
+}
```

Comment:
[question] this type can take a few different forms I think - an empty object or one with at least some of these properties (e.g. `showBranches`, `showRegression`). Is this expressable in TS?

Reply:
All the properties are optional. Do you mean certain sets of properties always appear together? There might be a way with [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates).

Reply:
IMO, if some of the options have to co-occur with others, defining a type for each validation combination and then combining them in a union type would be the way to go.

Reply:
Trying [mapped types](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html) in https://github.com/nextstrain/auspice/pull/1864#discussion_r1828775085

Reply:
Similar to https://github.com/nextstrain/auspice/pull/1864#discussion_r1830060286 let's continue this discussion in a separate PR. I do think reducing the number of optional properties is going to be very powerful, but let's keep this PR as-is.

Reply:
Added a post-merge task:

> - Post-merge: consider using mapped types in a new PR

Reply:
@genehack re: union

> defining a type for each validation combination and then combining them in a union type would be the way to go

This doesn't work. Example:

```ts
interface RadialPhyloNode {
  x: number
  y: number
  angle: number
  // …
}

interface ScatterPhyloNode {
  x: string
  y: string
  // ...
}

export interface PhyloNode extends RadialPhyloNode, ScatterPhyloNode
```

There is a type error:

> Interface 'PhyloNode' cannot simultaneously extend types 'RadialPhyloNode' and 'ScatterPhyloNode'.
> Named property 'x' of types 'RadialPhyloNode' and 'ScatterPhyloNode' are not identical.

Even if this did work, it's disconnected from the underlying conditions (layout) and prone to error. We'd only be using the combined type which makes everything available, and there is no validation that `angle` is only used for radial layout.